Articles/Blog/2022-08-brightspot-ci.md

```diff
@@ -0,0 +1,93 @@
+
+Paul Romano noted it was surprisingly easy for the team to get a first useful CI pipeline working with Travis CI. Switching over the GitHub Actions was also relatively straightforward.
+
+One of the unique aspects of the project is that the Monte Carlo method, being stochastic in nature, results in the possibility of otherwise identical simulations diverging from one another on two systems with ever-so-slightly different characteristics; for example, a different implementation of standard library math functions might result in completely different answers (within uncertainty) if the simulations are run long enough.
+
+This has sometimes made it difficult to reproduce CI failures locally.  Romano has seen quite a few circumstances where developers report “it works for me but always fails on Travis/GitHub Actions”.  This is especially prone to occur when someone makes a change that *is* expected to change reference answers (against which the tests check).  Then, updating the reference results requires some sensitivity to the platform differences between development and CI environments to get just right.
+
+Another hidden challenge the team noticed is keeping up with changes in dependencies. Every so often, a library will change their API, deprecate something, or otherwise change some behavior that forces us to deal with it right away because not doing so means our tests no longer pass. ExaSMR tries to stay on top of this by *not* pinning dependencies to specific versions.  The consequence, however, is that issues pop up periodically and divert attention from other areas.
+
+### Visit
+
+The biggest testing focus is on the (C++) source code for [Visit](https://github.com/visit-dav/visit)’s executable metadata-server, engine, viewer and controller.  It has dozens of external dependencies for geometry, graphics, file formats, and I/O.  Visit’s code base was hosted on NERSC’s Subversion servers until early 2019 when it was [migrated](https://bssw.io/blog_posts/continuous-technology-refreshment-an-introduction-using-recent-tech-refresh-experiences-on-visit) to git and GitHub. [Cyrus Harrison](https://github.com/cyrush) first added CI to the codebase around March 2019. Prior to that, only [nightly testing](https://visit-dav.github.io/visit-website/dashboard/) was in use. Adding CI testing was complicated because of the graphical nature of the code and the large number of dependencies.  He opted to use docker images to provide pre-built dependency binaries, combined with a set of scripts that allowed running on both Travis and Circle-CI.  The process to develop the scripts took weeks of trial and error.  The team went on to add more features and switch to Azure pipelines in Aug. 2020.  The CI tests run using a vanilla RedHat Linux base image, but the release process also builds images for Centos, Fedora, Debian and Ubuntu.
```

Comment:
@frobnitzem, Cyrus Harrison is the only person named in this article who has a link.  Do we need that?